File: openclaw-ios/OpenClaw/Services/HybridVoiceService.swift
import SwiftUI
import Speech
import AVFoundation

/**
 * Hybrid Voice Service für iOS
 * Nutzt Siri für einfache Anfragen (kostenlos)
 * OpenClaw Backend nur für komplexe Anfragen
 */
class HybridVoiceService: ObservableObject {
    @Published var voiceState: VoiceState = .idle
    
    private var speechRecognizer: SFSpeechRecognizer?
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?
    private let audioEngine = AVAudioEngine()
    private let synthesizer = AVSpeechSynthesizer()
    
    enum VoiceState {
        case idle, listening, processing, speaking, error
    }
    
    enum ResultSource {
        case builtinSiri    // Kostenlos
        case externalAI     // Bezahlt
    }
    
    struct HybridResult {
        let source: ResultSource
        let content: String?
        let siriIntent: SiriIntent?
        let estimatedCost: String
    }
    
    struct SiriIntent {
        let intent: String
        let parameters: [String: Any]
    }
    
    init() {
        speechRecognizer = SFSpeechRecognizer(locale: Locale(identifier: "de-DE"))
        requestAuthorization()
    }
    
    private func requestAuthorization() {
        SFSpeechRecognizer.requestAuthorization { _ in }
        AVAudioSession.sharedInstance().requestRecordPermission { _ in }
    }
    
    /**
     * Startet Spracheingabe und entscheidet: Siri oder OpenClaw
     */
    func startVoiceInput(completion: @escaping (HybridResult) -> Void) {
        voiceState = .listening
        
        // Starte Spracherkennung
        startRecording { [weak self] transcript in
            self?.voiceState = .processing
            
            // Entscheide: Siri oder externe AI
            if self?.isBuiltInQuery(transcript) == true {
                // Nutze Siri (KOSTENLOS)
                self?.delegateToSiri(query: transcript)
                completion(HybridResult(
                    source: .builtinSiri,
                    content: nil,
                    siriIntent: SiriIntent(intent: "custom", parameters: ["query": transcript]),
                    estimatedCost: "Kostenlos"
                ))
            } else {
                // Sende an OpenClaw Backend
                completion(HybridResult(
                    source: .externalAI,
                    content: nil,
                    siriIntent: nil,
                    estimatedCost: "~$0.003"
                ))
            }
        }
    }
    
    /**
     * Prüft ob Siri die Anfrage beantworten kann
     */
    private func isBuiltInQuery(_ query: String) -> Bool {
        let lower = query.lowercased()
        
        let siriPatterns = [
            /^(wie spät|uhrzeit|wie viel uhr)/,
            /^(welcher tag|datum|wann ist heute)/,
            /^(wetter|temperatur|regnet es)/,
            /^(timer|wecker|erinner mich|termin)/,
            /^(navigiere|weg nach|route zu)/,
            /^(wer ist|was ist|wo liegt)/,
            /^(rechne|wie viel ist \d)/,
            /^(licht|lampe|musik|lautstärke)/,
            /^(ruf|anrufen|nachricht|sms)/
        ]
        
        return siriPatterns.contains { lower.range(of: $0, options: .regularExpression) != nil }
    }
    
    /**
     * Delegiert an Siri (kostenlos)
     */
    private func delegateToSiri(query: String) {
        // Siri wird die Anfrage nativ beantworten
        // Keine externe API nötig
        voiceState = .idle
    }
    
    /**
     * Lokale Spracherkennung
     */
    private func startRecording(completion: @escaping (String) -> Void) {
        // Implementation der Speech Recognition
        // ...
    }
    
    /**
     * Text-to-Speech
     */
    func speak(_ text: String) {
        voiceState = .speaking
        
        let utterance = AVSpeechUtterance(string: text)
        utterance.voice = AVSpeechSynthesisVoice(language: "de-DE")
        utterance.rate = 0.5
        
        synthesizer.speak(utterance)
    }
    
    func stop() {
        audioEngine.stop()
        recognitionRequest?.endAudio()
        voiceState = .idle
    }
}

/**
 * SwiftUI View für Voice Interface mit Kostenschätzung
 */
struct HybridVoiceButton: View {
    @StateObject private var voiceService = HybridVoiceService()
    @State private var showCostInfo = false
    
    var body: some View {
        VStack {
            Button(action: {
                voiceService.startVoiceInput { result in
                    handleResult(result)
                }
            }) {
                Image(systemName: voiceIcon)
                    .font(.system(size: 30))
                    .foregroundColor(voiceColor)
                    .frame(width: 70, height: 70)
                    .background(voiceColor.opacity(0.2))
                    .clipShape(Circle())
                    .scaleEffect(voiceService.voiceState == .listening ? 1.2 : 1.0)
                    .animation(.easeInOut(duration: 0.3), value: voiceService.voiceState)
            }
            
            if showCostInfo {
                Text("Kostenlos über Siri")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
        }
    }
    
    private var voiceIcon: String {
        switch voiceService.voiceState {
        case .idle: return "mic.fill"
        case .listening: return "waveform"
        case .processing: return "brain"
        case .speaking: return "speaker.wave.2.fill"
        case .error: return "exclamationmark.triangle.fill"
        }
    }
    
    private var voiceColor: Color {
        switch voiceService.voiceState {
        case .idle: return .gray
        case .listening: return .accentColor
        case .processing: return .orange
        case .speaking: return .green
        case .error: return .red
        }
    }
    
    private func handleResult(_ result: HybridVoiceService.HybridResult) {
        showCostInfo = (result.source == .builtinSiri)
        
        // Verberge Cost Info nach 3 Sekunden
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
            showCostInfo = false
        }
    }
}

File: openclaw-android/app/src/main/java/de/openclaw/assistant/service/HybridVoiceService.kt
package de.openclaw.assistant.service

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import android.speech.tts.TextToSpeech
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.Locale

/**
 * Hybrid Voice Service
 * Nutzt Google Assistant für einfache Anfragen (kostenlos)
 * OpenClaw Backend nur für komplexe Anfragen
 */
class HybridVoiceService(
    private val activity: Activity,
    private val onQueryResult: (HybridResult) -> Unit
) {
    private val _voiceState = MutableStateFlow(VoiceState.IDLE)
    val voiceState: StateFlow<VoiceState> = _voiceState

    private var textToSpeech: TextToSpeech? = null
    private var speechRecognizer: SpeechRecognizer? = null

    enum class VoiceState {
        IDLE, LISTENING, PROCESSING, SPEAKING, ERROR
    }

    enum class ResultSource {
        BUILTIN_GOOGLE, // Kostenlos - Google Assistant
        EXTERNAL_AI     // Bezahlt - OpenClaw Backend
    }

    data class HybridResult(
        val source: ResultSource,
        val content: String?,
        val action: AssistantAction?,
        val estimatedCost: String
    )

    data class AssistantAction(
        val type: String,
        val data: Map<String, Any>
    )

    init {
        initTextToSpeech()
    }

    private fun initTextToSpeech() {
        textToSpeech = TextToSpeech(activity) { status ->
            if (status == TextToSpeech.SUCCESS) {
                textToSpeech?.language = Locale.GERMAN
            }
        }
    }

    /**
     * Startet Spracheingabe
     * Entscheidet automatisch: Google Assistant oder OpenClaw
     */
    fun startVoiceInput() {
        _voiceState.value = VoiceState.LISTENING

        // Option 1: Google Assistant direkt (für einfache Anfragen)
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.GERMAN)
            putExtra(RecognizerIntent.EXTRA_PROMPT, "Sprich zu OpenClaw...")
        }

        try {
            activity.startActivityForResult(intent, REQUEST_VOICE_INPUT)
        } catch (e: ActivityNotFoundException) {
            // Fallback zu lokalem SpeechRecognizer
            startLocalSpeechRecognition()
        }
    }

    /**
     * Verarbeitet die Spracheingabe
     * Entscheidet: Google Assistant oder externe AI
     */
    fun processVoiceInput(transcript: String) {
        _voiceState.value = VoiceState.PROCESSING

        // Check: Ist das eine einfache Anfrage für Google Assistant?
        when {
            isBuiltInQuery(transcript) -> {
                // Leite an Google Assistant weiter (KOSTENLOS)
                Log.d(TAG, "Routing to Google Assistant (FREE): $transcript")
                delegateToGoogleAssistant(transcript)
            }
            else -> {
                // Sende an OpenClaw Backend (mit Kosten)
                Log.d(TAG, "Routing to OpenClaw Backend: $transcript")
                onQueryResult(HybridResult(
                    source = ResultSource.EXTERNAL_AI,
                    content = null, // Wird vom Backend gefüllt
                    action = null,
                    estimatedCost = "~$0.003"
                ))
            }
        }
    }

    /**
     * Prüft ob Google Assistant die Anfrage beantworten kann
     */
    private fun isBuiltInQuery(query: String): Boolean {
        val lower = query.lowercase()

        val builtInPatterns = listOf(
            Regex("^(wie spät|wieviel uhr|uhrzeit)"),
            Regex("^(welcher tag|datum|wann ist heute)"),
            Regex("^(wetter|temperatur|regnet es)"),
            Regex("^(timer|wecker|erinner mich|termin)"),
            Regex("^(navigiere|weg nach|route zu)"),
            Regex("^(wer ist|was ist|wo liegt|wie groß)"),
            Regex("^(rechne|wie viel ist [0-9])"),
            Regex("^(licht|lampe|musik|lautstärke)")
        )

        return builtInPatterns.any { it.containsMatchIn(lower) }
    }

    /**
     * Delegiert an Google Assistant (kostenlos)
     */
    private fun delegateToGoogleAssistant(query: String) {
        val intent = Intent(Intent.ACTION_VOICE_COMMAND).apply {
            putExtra(Intent.EXTRA_TEXT, query)
        }

        try {
            activity.startActivity(intent)
            onQueryResult(HybridResult(
                source = ResultSource.BUILTIN_GOOGLE,
                content = null,
                action = AssistantAction("delegate_to_google", mapOf("query" to query)),
                estimatedCost = "Kostenlos"
            ))
        } catch (e: ActivityNotFoundException) {
            // Fallback: Zeige Google Suche
            val searchIntent = Intent(Intent.ACTION_WEB_SEARCH).apply {
                putExtra(SearchManager.QUERY, query)
            }
            activity.startActivity(searchIntent)
        }

        _voiceState.value = VoiceState.IDLE
    }

    /**
     * Lokale Spracherkennung (Fallback)
     */
    private fun startLocalSpeechRecognition() {
        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(activity)
        // ... Implementation
    }

    /**
     * Text-to-Speech für Antworten
     */
    fun speak(text: String) {
        _voiceState.value = VoiceState.SPEAKING
        textToSpeech?.speak(text, TextToSpeech.QUEUE_FLUSH, null, null)
    }

    fun stop() {
        textToSpeech?.stop()
        speechRecognizer?.destroy()
        _voiceState.value = VoiceState.IDLE
    }

    companion object {
        private const val TAG = "HybridVoiceService"
        const val REQUEST_VOICE_INPUT = 1001
    }
}
